keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,389 @@
+  instance
+- Bundle all diagnostics artifacts into an archive
+
+### Non-functional
+
+1.  Provide an **easy** experience for operator users to collect diagnostic
+    artifact archives
+2.  Be resilient to faults and failures. Collect as much diagnostics artifacts
+    as possible and allow failed collections to be retried (idempotency) and
+    incremented to (like `wget`'s `--continue` flag), in a way that collection
+    is _resumable_
+3.  Incorporate standard tools that are already provided by either organizations
+    or the community behind applications as much as possible
+
+## Non-goals
+
+- Extensive collection of Kubernetes-related diagnostics artifacts
+- At least not initially: collection of metrics from monitoring services (e.g.,
+  Prometheus, Statsd, etc.).
+- Automatic fixing of faults
+- Preflight checks
+- Analysis of collected artifacts
+- Extending diagnostics bundle with custom artifact collectors
+
+## Requirements
+
+- MUST create an archive with diagnostics artifacts related specifically to an
+  operator instance
+- MUST include application-related diagnostics artifacts in the archive
+- MUST include instance-related diagnostics artifacts in the archive
+- MUST include KUDO-related diagnostics artifacts in the archive
+- MUST include "Kubernetes workload"-related diagnostics artifacts in the
+  archive
+- MUST accept parameters and work without interactive prompts
+- SHOULD work in airgapped environments
+- SHOULD report the versions of every component and tool, in the archive (e.g.,
+  the version of the collector, the application version, the operator version,
+  the KUDO version, the Kubernetes version, etc.)
+- SHOULD follow Kubernetes' ecosystem conventions and best practices
+- MUST be published as a static binary
+- SHOULD make it possible to publish archive to cloud object storage (AWS S3,
+  etc.)
+- MUST follow SemVer
+
+## Proposal
+
+### Operator user experience
+
+The output from diagnostics collection is an archive containing all
+diagnostics artifacts for the provided operator instance.
+
+```bash
+kubectl kudo diagnostics collect --instance=%instance% --namespace=%namespace%
+```
+
+### Operator developer experience
+
+To configure diagnostics globally, this KEP introduces an optional top-level
+`diagnostics` key in operator.yaml.
+
+#### Diagnostics collection
+
+The following diagnostics will be implicitly collected without any configuration
+from the operator developer:
+
+- Logs for deployed pods related to the KUDO Instance
+- YAML for created resources, including both spec and status, related
+  to the KUDO instance
+- Output of `kubectl describe` for all deployed resources related to the KUDO
+  Instance
+- Current plan status, if one exists, or the KUDO Instance
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
```

Comment:
I think it's debatable whether we should use `describe` or just a `get -o yaml`... I'd vote for the latter since it's far more machine-readable.

Reply:
We could also do both _yaml_ and _describe_, as we anyway seem to plan to do it both ways for the KUDO instance resources